### PR TITLE
Activiti-3310 Variable Mapping Logic

### PR DESCRIPTION
**Pull request by miguelruizdev**
Includes changes relative to the enablement of variable mapping for user tasks and call activities
Issues tracked by JIRA: 
https://issues.alfresco.com/jira/browse/ACTIVITI-3310
https://issues.alfresco.com/jira/browse/ACTIVITI-3295
https://issues.alfresco.com/jira/browse/ACTIVITI-3240

### CODE REVIEW

**Comment by CTI777** [comment]
There are still some open questions:
1. In the case when no mapping is defined for the task in process extension file - we map all variables (like before). We do not have a possibility to not allow variable mapping for all process variables (input / output).
2. Another question to keep in mind is that we may have a case (currently) when actual processVariables are different from the processVariables defined in the process extension file. There is a possibility to add process variables (from API or using activiti core functions directly). Routines for adding process variables not always check processExtension file...

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/DefaultActivityBehaviorFactoryMapping.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    
+
+    @Override
+    public UserTaskActivityBehavior createUserTaskActivityBehavior(UserTask userTask) {
+        return new DefaultUserTaskBehavior(userTask,
+                                           new VariablesMappingProvider(processExtensionService));
+    }
+
+
```
Comment (eromano) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/DefaultActivityBehaviorFactoryMapping.java]: do you need three spaces here?

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/DefaultActivityBehaviorFactoryMapping.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public class DefaultActivityBehaviorFactoryMapping extends DefaultActivityBehaviorFactory {
+
+    private ProcessExtensionService processExtensionService;
+
+    public DefaultActivityBehaviorFactoryMapping(ProcessExtensionService processExtensionService) {
+        super();
+        this.processExtensionService = processExtensionService;
+    }
+
```
Comment (eromano) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/DefaultActivityBehaviorFactoryMapping.java]: double space

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/DefaultUserTaskBehavior.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+public class DefaultUserTaskBehavior extends UserTaskActivityBehavior {
+
+    VariablesMappingProvider mappingProvider;
+  
+    public DefaultUserTaskBehavior(UserTask userTask,
+                                   VariablesMappingProvider mappingProvider) {
+        super(userTask);
+        this.mappingProvider = mappingProvider;
+    }
+
```
Comment (eromano) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/DefaultUserTaskBehavior.java]: double space

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/DefaultUserTaskBehavior.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    protected Map<String, Object> getOutBoundVariables(CommandContext commandContext,
+                                                       DelegateExecution execution, 
+                                                       Map<String, Object> taskCompleteVariables) {
+        
+        return mappingProvider.calculateOutPutVariables(commandContext.getProcessEngineConfiguration().isCopyVariablesToLocalForTasks(),
+                                                        execution.getProcessDefinitionId(), 
+                                                        execution.getCurrentActivityId(),
+                                                        taskCompleteVariables); 
+    }
+
```
Comment (eromano) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/DefaultUserTaskBehavior.java]: threee space

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                    if (Mapping.SourceMappingType.STATIC_VALUE.equals(inputMapping.getType())) {
+                        return inputMapping.getValue();         
+                    } 
+                }
+                
+            }
+        }
+        return null;
+    }
+
```
Comment (eromano) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java]: two spaces

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                //We have to check if this is needed?
+                else {
+                    if (Mapping.SourceMappingType.STATIC_VALUE.equals(inputMapping.getType())) {
+                        return inputMapping.getValue();         
+                    } 
+                }
+                
+            }
+        }
+        return null;
```
Comment (eromano) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java]: not propagate null

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    }
+    
+    
+    public Map<String, Object> calculateInputVariables(DelegateExecution execution) {
+        
+        Map<String, Object> inboundVariables = null;
+        ProcessExtensionModel extensions = processExtensionService.getExtensionsForId(execution.getProcessDefinitionId());      
+        if (extensions != null) {
+            ProcessVariablesMapping processVariablesMapping = extensions.getExtensions().getMappingForFlowElement(execution.getCurrentActivityId());
```
Comment (eromano) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java]: to dense

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+                    String name = mapping.getValue().toString();
+                    
+                    return activitiCompleteVariables != null ?
+                           activitiCompleteVariables.get(name) :
+                           null;     
+                }
+            }
+            
+        }
+        return null;
```
Comment (eromano) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java]: not propagate null

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    
+    public Map<String, Object> calculateOutPutVariables(boolean defaultCopyAllVariables,
+                                                        String processDefinitionId, 
+                                                        String activityId,
+                                                        Map<String, Object> activitiCompleteVariables) {
+        
+        Map<String, Object> outboundVariables = new HashMap<>();    
+        
+        if (activitiCompleteVariables != null && !activitiCompleteVariables.isEmpty()) {
+            ProcessExtensionModel extensions = processExtensionService.getExtensionsForId(processDefinitionId);
```
Comment (eromano) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/VariablesMappingProvider.java]: to dense

File: activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/UserTaskActivityBehavior.java [hunk 10]
```diff
@@ -235,15 +234,68 @@ public void execute(DelegateExecution execution) {
...
+              TaskVariableCopier.copyVariablesIntoTaskLocal(task);
+          }     
+      } else {
+          //Check what to do if empty mapping
+          task.setVariablesLocal(inboundVars);       
+      }    
+  }
+  
+  protected Map<String, Object> getInboundVariables(DelegateExecution execution) {
+      return null;
```
Comment (eromano) [inline activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/UserTaskActivityBehavior.java]: a method that just returns null?

File: activiti-engine/src/main/java/org/activiti/engine/impl/cmd/CompleteTaskCmd.java [hunk 11]
```diff
@@ -62,11 +62,16 @@ protected Void execute(CommandContext commandContext, TaskEntity task) {
         task.setTransientVariables(transientVariables);
       }
     }
-
-    if(commandContext.getProcessEngineConfiguration().isCopyVariablesToLocalForTasks()){
-      TaskVariableCopier.copyVariablesOutFromTaskLocal(task);
-    }
-
+//    We have moved this logic to UserTaskActivityBehavior.trigger(..
```
Comment (eromano) [inline activiti-engine/src/main/java/org/activiti/engine/impl/cmd/CompleteTaskCmd.java]: commented code do we need it?

File: activiti-engine/src/main/java/org/activiti/engine/impl/cmd/CompleteTaskCmd.java [hunk 12]
```diff
@@ -76,4 +81,21 @@ protected String getSuspendedTaskException() {
...
+  }
+  
+  public void setTaskVariables(Map<String, Object> taskVariables) {
+      this.taskVariables = taskVariables;  
+  }
+  
+  public Map<String, Object> getTaskVariables() {
+      return taskVariables;  
+  }
+
```
Comment (eromano) [inline activiti-engine/src/main/java/org/activiti/engine/impl/cmd/CompleteTaskCmd.java]: double space

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+                .extracting(VariableInstance::getName,
+                            VariableInstance::getValue)
+                .containsOnly(tuple("task-variable-name-1",
+                                    "value1"),
+                              tuple("task-variable-name-2",
+                                    0),
+                              tuple("task-variable-name-3",
+                                    5)
+                              );
+
```
Comment (eromano) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java]: three space

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+        //Complete task with variables
+        //This will complete also our process
+        Map<String, Object> variablesForTask = new HashMap<>();
+        variablesForTask.put("task-variable-name-1", "value from subprocess");
+        variablesForTask.put("task-variable-name-2", 39);
+        
+        completeTask(task.getId(), variablesForTask);
+ 
+    }
+
```
Comment (eromano) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java]: three space

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+        assertThat(subProcessInstance.getProcessDefinitionKey()).isEqualTo(SUB_PROCESS_CALL_ACTIVITY);
+        
+        
+        //Check SubProcess Variables
+        procVariables = processRuntime.variables( ProcessPayloadBuilder
+                                                 .variables()
+                                                 .withProcessInstanceId(subProcessInstance.getId())
+                                                 .build());
+        
+       //This we expect to have after implementation of mapping
```
Comment (eromano) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java]: commented code do we need it?

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+        assertThat("my-task-call-activity").isEqualTo(task.getName());
+        
+        
+        //Check Task Variables: only variables defined in 'mappings' should be present
+        List<VariableInstance> taskVariables = taskRuntime.variables(TaskPayloadBuilder
+                                                                         .variables()
+                                                                         .withTaskId(task.getId())
+                                                                         .build());
+        
+        //This should work after implementation
```
Comment (eromano) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java]: commented code do we need it?

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+                              "inName"),
+                        tuple("out-unmapped-variable-matching-name",
+                              "default"),
+                        tuple("input-unmapped-variable-with-matching-name",
+                              "inTest"),
+                        tuple("age",
+                               20)
+                );
+         
+        //Check userTask is available
```
Comment (eromano) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeCallActivityMappingIT.java]: commented code do we need it?

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskRuntimeVariableMappingTest.java [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+                      tuple("process_variable_inputmap_1",
+                            "inputmap1Value"),          //Should be unchanged
+                      tuple("process_variable_outputmap_1",
+                            "outputTaskValue")          //Should be changed to 'outputTaskValue' after implementation of mapping
+                                                      
+        );    
+                
+        processRuntime.delete(ProcessPayloadBuilder.delete().withProcessInstance(process).build());
+    }
+
```
Comment (eromano) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskRuntimeVariableMappingTest.java]: three space

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskRuntimeVariableMappingTest.java [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+                        tuple("process_variable_inputmap_1",
+                                "inputmap1Value"),
+                        tuple("process_variable_outputmap_1",
+                                "outputmap1Value"));
+
+
+        processRuntime.delete(ProcessPayloadBuilder.delete().withProcessInstance(process).build());
+
+    }
+
```
Comment (eromano) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskRuntimeVariableMappingTest.java]: three space

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskRuntimeVariableMappingTest.java [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+                .isNotNull()
+                .extracting(VariableInstance::getName,
+                        VariableInstance::getValue)
+                .containsOnly(tuple("process_variable_unmapped_1",
+                        "unmapped1Value"),
+                        tuple("process_variable_inputmap_1",
+                                "inputmap1Value"),
+                        tuple("process_variable_outputmap_1",
+                                "outputmap1Value"));
+
```
Comment (eromano) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskRuntimeVariableMappingTest.java]: two space

**Comment by eromano** [comment]
@miguelruizdev https://app.codacy.com/app/Activiti/Activiti/pullRequest?prid=3808582 is highlighting some possibble improvements too

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskRuntimeVariableMappingTest.java [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+        securityUtil.logInAs("garth");
+        ProcessInstance process = processRuntime.start(ProcessPayloadBuilder.start()
+                                                               .withProcessDefinitionKey(TASK_VAR_NO_MAPPING)
+                                                               .build());
+
+        Page<Task> tasks = taskRuntime.tasks(Pageable.of(0,
+                                                         50));
+
+        assertThat(tasks.getContent()).hasSize(1);
+        Task task = tasks.getContent().get(0);
```
Comment (erdemedeiros) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskRuntimeVariableMappingTest.java]: Block from line 163 to 167 can be extracted in a method and reused across the different tests.

File: activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/ProcessVariablesInitiator.java [hunk 22]
```diff
@@ -45,21 +45,39 @@ public ProcessVariablesInitiator(ProcessExtensionService processExtensionService
         this.variableValidationService = variableValidationService;
     }
 
-    @Override
-    public ProcessInstance createAndStartProcessInstanceWithInitialFlowElement(ProcessDefinition processDefinition, String businessKey, String processInstanceName, FlowElement initialFlowElement, Process process, Map<String, Object> variables, Map<String, Object> transientVariables, boolean startProcessInstance) {
+    public Map<String, Object> getVariablesFromExtensionFile(ProcessDefinition processDefinition,
```
Comment (erdemedeiros) [inline activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/ProcessVariablesInitiator.java]: Can we rename it to `calculateVariablesFromExtensionFile`?

File: activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/ProcessVariablesInitiator.java [hunk 23]
```diff
@@ -45,21 +45,39 @@ public ProcessVariablesInitiator(ProcessExtensionService processExtensionService
...
+        }
+
+        return processedVariables;
+    }
+
+
+    @Override
+    public ProcessInstance createAndStartProcessInstanceWithInitialFlowElement(ProcessDefinition processDefinition, String businessKey, String processInstanceName, FlowElement initialFlowElement, Process process, Map<String, Object> variables, Map<String, Object> transientVariables, boolean startProcessInstance) {
+        if (processExtensionService.hasExtensionsFor(processDefinition)) {
+            Map<String, Object> processedVariables=getVariablesFromExtensionFile(processDefinition,variables);
```
Comment (erdemedeiros) [inline activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/ProcessVariablesInitiator.java]: Please, use spaces before and after `=` operator.

File: activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/model/Extension.java [hunk 24]
```diff
@@ -61,4 +61,20 @@ public VariableDefinition getPropertyByName(String name){
        
         return null;
     }
+
+    public boolean hasEmptyInputsMapping(String elementId){
+        ProcessVariablesMapping processVariablesMapping = mappings.get(elementId);
+        return processVariablesMapping != null && processVariablesMapping.getInputs().size() == 0;
+    }
+
+    public boolean hasEmptyOututsMapping(String elementId){
```
Comment (erdemedeiros) [inline activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/model/Extension.java]: Typo.

File: activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/model/Extension.java [hunk 25]
```diff
@@ -61,4 +61,20 @@ public VariableDefinition getPropertyByName(String name){
        
         return null;
     }
+
+    public boolean hasEmptyInputsMapping(String elementId){
```
Comment (erdemedeiros) [inline activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/model/Extension.java]: Please cover these 3 methods with unit tests.

File: activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/ProcessVariablesInitiator.java (hunk 22 above)
Comment (almerico) [reply activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/ProcessVariablesInitiator.java]: ok

File: activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/model/Extension.java (hunk 24 above)
Comment (almerico) [reply activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/model/Extension.java]: ok

File: activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/CallActivityBehavior.java [hunk 26]
```diff
@@ -204,4 +216,26 @@ public void setProcessDefinitonKey(String processDefinitonKey) {
...
+  }
+
+
+  protected Map<String, Object> getOutBoundVariables(CommandContext commandContext,
+                                                     DelegateExecution execution,
+                                                     Map<String, Object> taskVariables) {
+    return new HashMap<String, Object>();
+  }
+
+  protected Map<String, Object> getVariablesFromExtensionFile(ProcessDefinition processDefinition) {
```
[the remark this answers is not in the document]
Comment (almerico) [reply activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/CallActivityBehavior.java]: do not understand proposed change

[the remark this answers is not in the document]
Comment (almerico) [reply activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/CallActivityBehavior.java]: rename proposed?

[the remark this answers is not in the document]
Comment (miguelruizdev) [reply activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/CallActivityBehavior.java]: the logic added in called activity behavior related to **getVariablesFromExtensionFile** needs to be addressed in the subclass created for the new behavior.
Everything related to it needs to be passed to **MappingAwareCallActivityBehavior**

**Pull request merged**